### ISSUE TITLE
Not finishing sync and intermittent crashes

### ISSUE THREAD

**Issue by sidhujag** [syscoin/rosetta-syscoin#1]
The integration from bitcoin was pretty simple mostly parameterization and renaming. There wasn't any specific changes I needed to make, the tests pass but CLI obviously doesn't because I guess the rosetta node isn't synced to the tip and I cannot get it to sync to the tip.

After the node fully syncs from scratch and then rosetta node indexes the blocks:

`2020/11/19 20:52:23 Syncing 612569-772333`

The blocks around 612k were have few thousand transactions per block (1 min block time) for a bunch of blocks. It seemed to get stuck here.

After a while the rosetta node just shuts down with no error message. I am running the docker without -d so I can see the output of the docker in ubuntu 18.04.

Might be related to garbage collection and/or pruning. I see the syscoin node getting new blocks but the rosetta node isn't "catching up"

`2020-11-19T23:04:04.392Z	DEBUG	memory	utils/utils.go:62	stats	{"heap (MB)": 9110.803466796875, "max heap (MB)": 9110.803466796875, "stack (MB)": 439.34375, "system (MB)": 10668.136474609375, "garbage collections": 13}
2020-11-19T23:04:19.224Z	DEBUG	syscoind	syscoin/node.go:55	UpdateTip: new best=140adc5b205ceb60cc5e1a89728d99f2c46fbd484fce9c29b84ac10217992283 height=772444 version=0x10000104 log2_work=89.192392 tx=1596590 date='2020-11-19T23:03:31Z' progress=1.000000 cache=0.1MiB(381txo) evodb_cache=0.0MiB
2020-11-19T23:04:43.173Z	DEBUG	syscoind	syscoin/node.go:55	UpdateTip: new best=5d91aea1e61511114783d17e6916af865ec80bd981d065c6f474fc0a4cf5089b height=772445 version=0x10000104 log2_work=89.192395 tx=1596591 date='2020-11-19T23:04:09Z' progress=1.000000 cache=0.1MiB(383txo) evodb_cache=0.0MiB
2020-11-19T23:06:25.774Z	DEBUG	syscoind	syscoin/node.go:55	UpdateTip: new best=0c231e6306feb2c1c1b32b1da2eae73589b5d04632fe7249b80475b6d1ec0fdd height=772446 version=0x10000104 log2_work=89.192398 tx=1596592 date='2020-11-19T23:04:32Z' progress=0.999999 cache=0.1MiB(385txo) evodb_cache=0.0MiB
2020-11-19T23:06:25.774Z	DEBUG	syscoind	syscoin/node.go:55	UpdateTip: new best=26188c210a6d75191c092ed5cae8c5714b0200ff470e9cd36c89ab41c85189bd height=772447 version=0x10000104 log2_work=89.192401 tx=1596593 date='2020-11-19T23:05:40Z' progress=1.000000 cache=0.1MiB(387txo) evodb_cache=0.0MiB
2020-11-19T23:06:25.774Z	DEBUG	syscoind	syscoin/node.go:55	UpdateTip: new best=965a7bb28ccc263390997ea6c6b0581ec58e2612fdfa4fe2bf18617ad101f690 height=772448 version=0x10000104 log2_work=89.192404 tx=1596594 date='2020-11-19T23:06:10Z' progress=1.000000 cache=0.1MiB(389txo) evodb_cache=0.0MiB
2020-11-19T23:06:52.735Z	DEBUG	memory	utils/utils.go:62	stats	{"heap (MB)": 9440.363388061523, "max heap (MB)": 9440.374649047852, "stack (MB)": 615.59375, "system (MB)": 10734.20191192627, "garbage collections": 14}
2020-11-19T23:07:38.147Z	DEBUG	syscoind	syscoin/node.go:55	UpdateTip: new best=ae2199ccd59d079952bba5135797681b16965c745a47237f033f0690141f853e height=772449 version=0x10000104 log2_work=89.192406 tx=1596595 date='2020-11-19T23:06:21Z' progress=0.999999 cache=0.1MiB(392txo) evodb_cache=0.0MiB
2020-11-19T23:08:24.383Z	DEBUG	syscoind	syscoin/node.go:55	UpdateTip: new best=7253c94f8665091901f7ef6b9f25ae8d6b15c9514e536250c1eb954e9e87da01 height=772450 version=0x10000104 log2_work=89.192409 tx=1596597 date='2020-11-19T23:07:23Z' progress=0.999999 cache=0.1MiB(386txo) evodb_cache=0.0MiB
2020-11-19T23:09:29.488Z	DEBUG	memory	utils/utils.go:62	stats	{"heap (MB)": 9159.375259399414, "max heap (MB)": 9440.816246032715, "stack (MB)": 623.75, "system (MB)": 10734.20191192627, "garbage collections": 15}
2020-11-19T23:10:15.631Z	DEBUG	syscoind	syscoin/node.go:55	UpdateTip: new best=e29600c6cffb4907e200625446b43fec8eb70c10d53313579e5c12541dbe3659 height=772451 version=0x10000104 log2_work=89.192412 tx=1596598 date='2020-11-19T23:08:14Z' progress=0.999999 cache=0.1MiB(388txo) evodb_cache=0.0MiB
2020-11-19T23:10:15.631Z	DEBUG	syscoind	syscoin/node.go:55	UpdateTip: new best=c87be8bea7ba15d22273ffbe37a917006b40f13231c9aba482efca20a3ffee13 height=772452 version=0x10000104 log2_work=89.192415 tx=1596599 date='2020-11-19T23:09:28Z' progress=1.000000 cache=0.1MiB(390txo) evodb_cache=0.0MiB
2020-11-19T23:10:15.631Z	DEBUG	syscoind	syscoin/node.go:55	UpdateTip: new best=d1683c9abea22449e38d8d370fd26bb80845c6ac5b888807035b2a45b805a8d3 height=772453 version=0x10000104 log2_work=89.192418 tx=1596600 date='2020-11-19T23:09:58Z' progress=1.000000 cache=0.1MiB(392txo) evodb_cache=0.0MiB
jag@jag-VirtualBox:~/rosetta-syscoin$ 
`

**Comment by sidhujag** [comment]
Looks like its related to block weight > 4MB as Bitcoin supports up to 4mb this block is an offender https://chainz.cryptoid.info/sys/block.dws?612480.htm

It is 1.33MB and over 4MB block weight, could this be the cause? investigating further

**Comment by sidhujag** [comment]
It seems the addblocks calls UpdateBalance on the SDK which calls getHistoricalBalance which does a database scan for each account. This transaction has over 30k accounts so it ends up looking up over 30k accounts, it seems about half of the accounts are looked up before memory runs out 16GB on my virtual machine. So there seems to be some large overhead pertaining to looking up accounts on balances.

**Comment by sidhujag** [comment]
It is not strictly related to the size because the block 612509 is always passed it has nearly 40k operations and has no problem. The only difference is that 612569 has less transactions but larger amount of outputs (around 250 each) it has less total operations but more outputs per transaction maybe it triggers an issue in the balance update mechanism of the SDK due to the structure of the block.

**Comment by sidhujag** [comment]
Refactored the update balance calls on the sdk to use a database get via storing account balance inside of the account table, seems to run faster (5hours native vs 15 seconds to verify offending block with the fix)

**Comment by patrick-ogrady** [comment]
Thanks for doing all this analysis, @sidhujag! Here are my thoughts:

> It seems to stop on this block https://chainz.cryptoid.info/sys/block.dws?612569.htm. Either the garbage collection is deterministic and failing there or that block being 2.5MB causes an issue with the indexer.

The garbage collection in Golang is not deterministic and would bet all my money that the larger block here is causing the issue. This could be one of a few things:
* We store pre-processed data on disk (convert the raw blockchain data to Rosetta), so there is a small expansion factor when we write that can increase latency.
* DB lookup latency is exploding because of the number of concurrent queries we are performing during `AddBlock`.
* Your allocated RAM is low so DB operations are causing ["thrashing"]((https://en.wikipedia.org/wiki/Thrashing_(computer_science)).

> It seems the addblocks calls UpdateBalance on the SDK which calls getHistoricalBalance which does a database scan for each account. This transaction has over 30k accounts so it ends up looking up over 30k accounts (asynchronously which eats memory as results are deferred until the fetches are done), it seems about half of the accounts are looked up before memory runs out 16GB on my virtual machine. So there seems to be some large overhead pertaining to looking up accounts on balances.
If OOM is typical for badger maybe we need to consider throttling on it like so https://discuss.dgraph.io/t/badger-performance-and-memory-usage-recommendations/4297
But the weird thing is even with 100GB swap it still doesn't complete indexing the block which leads me to think there is either circular contention happening or some issue related to the number of operations and scan calls to the db in parallel.
I let it run for about an hour and it still did not sync that block, because of the swap it doesn't crash but it seems locked up in the db scan calls.

Performance will SIGNIFICANTLY degrade if the implementation is using swap extensively. This is related to a problem called ["trashing"](https://en.wikipedia.org/wiki/Thrashing_(computer_science)). To sync, you may need to increase the RAM you have allocated until we can find a way to reduce the memory pressure of adding a block. You could always switch down to a smaller box after you sync the block 😬 .

> It is not strictly related to the size because the block 612509 is always passed it has nearly 40k operations (over 3mb) and has no problem. The only difference is that 612569 has less transactions but larger amount of outputs (around 250 each) it has less total operations but more outputs per transaction maybe it triggers an issue in the balance update mechanism of the SDK due to the structure of the block.

In the `AddBlock` function, we also update the state of `CoinStorage` and `BalanceStorage` (in addition to actually storing the block). Updating the state of these stores (as you mentioned) can be quite intensive but is the only way to provide the endpoints we care about without a SQL backend (we need to denormalize and duplicate the data).

I've been working on a lot of optimizations on the `storage` layer for `rosetta-cli`. Once these are all merged, I can take another pass to see what I can improve in `rosetta-bitcoin`. If you are looking for results now, bumping up RAM should improve performance (as I mentioned earlier)! Fortunately, this occurs deterministically at this block so it should be easier to debug and fix.

> Refactored the update balance calls on the sdk to use a database get via storing account balance inside of the account table, seems to run faster (5hours native vs 15 seconds to verify offending block with the fix). Will PR upstream.

Interesting! [Badger claims that their scan operation is extremely fast](https://dgraph.io/blog/post/badger-lmdb-boltdb/), but it is possible something we are doing is really killing performance. Does this change you made make it impossible to fetch historical balances?

**Comment by sidhujag** [comment]
Hi @patrick-ogrady 
Yes I have resolved the bottleneck yes I agree reads should be faster I bet there is some contention happening. Basically you are doing scans on the db which is really slow due to locks so I made it work by removing the scan entirely and depending on balan

**Comment by sidhujag** [comment]
Hi @patrick-ogrady 

According to the benchmarks i am seeing consistent results of around 5k rw per second I refactored out the scan which is a bottleneck we should discuss because the reorg code will also have the same issue. 80k operations / 5000 its roughly around the time I’m seeing so I think it’s as fast as it’s going to get based on benchmarks.

**Comment by sidhujag** [comment]
I also noticed bumping up ram was required as I got the dreaded [txn too large issue ](https://github.com/dgraph-io/badger/issues/441), because you are processing all changes without breaking them down it causes that. I have bumped up to 1024mb MaxTableSize and 256mb ValueLogFileSize to get passed the error with logic that since we support 4mb wire vs 1mb wire on btc we should bump this by 4x. It does get around the problem but we should probably break the db transactions into chunks of say 10000 operations at a time or something, which will let memory clean up happen at the same time.
